### PR TITLE
executor: use git refs for snapshot read/write

### PR DESCRIPTION
**Pull request by bduffany**
:arrow_right: depends on https://github.com/buildbuddy-io/buildbuddy/pull/5115

* Use main git ref (pushed branch name) for reading and writing
* Use fallback git refs for reading only (PR base branch and repo default branch, if different from PR base branch)

**Related issues**: N/A

### CODE REVIEW

File: enterprise/server/remote_execution/containers/firecracker/firecracker_test.go [hunk 1]
```diff
@@ -2371,3 +2357,7 @@ func appendToLog(message string) *repb.Command {
 			`},
 	}
 }
+
+func recyclingEnabledTask() *repb.ScheduledTask {
```
Comment (tylerwilliams) [inline enterprise/server/remote_execution/containers/firecracker/firecracker_test.go:2361]: eh?

File: enterprise/server/remote_execution/snaploader/snaploader.go [hunk 2]
```diff
@@ -182,24 +208,45 @@ func New(env environment.Env) (*FileCacheLoader, error) {
...
+	for i, key := range allKeys {
+		manifest, err := l.getSnapshot(ctx, key)
+		if err != nil {
+			lastErr = err
+			continue
+		}
+		if i == 0 {
+			log.CtxInfof(ctx, "Found preferred snapshot ref=%q in cache", key.GetRef())
+		} else {
+			log.CtxInfof(ctx, "Found fallback snapshot ref=%q in cache", key.GetRef())
```
Comment (tylerwilliams) [inline enterprise/server/remote_execution/snaploader/snaploader.go:223]: maybe our refs proto should just be a list of refs instead of a preferred + list of refs?

File: enterprise/server/remote_execution/snaploader/snaploader.go [hunk 3]
```diff
@@ -88,10 +104,19 @@ func fileDigest(filePath string) (*repb.Digest, error) {
 
 // Snapshot holds a snapshot manifest along with the corresponding cache key.
 type Snapshot struct {
-	key      *fcpb.SnapshotKey
+	writeKey *fcpb.SnapshotKey
+	readKey  *fcpb.SnapshotKey
 	manifest *fcpb.SnapshotManifest
 }
 
+func (s *Snapshot) GetWriteKey() *fcpb.SnapshotKey {
```
Comment (tylerwilliams) [inline enterprise/server/remote_execution/snaploader/snaploader.go:112]: I don't see this used anywhere yet, that intentional?

File: enterprise/server/remote_execution/snaploader/snaploader.go [hunk 4]
```diff
@@ -42,21 +42,33 @@ const (
...
+// SnapshotKeys returns the cache keys for potential snapshot matches,
+// as well as the key that should be written to.
+//
 // TODO: include a version number in the key somehow, so that
 // if we make breaking changes e.g. to the vmexec API or firecracker
 // version etc., we can ensure that incompatible snapshots don't get reused.
-func NewKey(task *repb.ExecutionTask, configurationHash, runnerID string) (*fcpb.SnapshotKey, error) {
-	pd, err := digest.ComputeForMessage(task.GetCommand().GetPlatform(), repb.DigestFunction_SHA256)
+func SnapshotKeys(task *repb.ScheduledTask, configurationHash, runnerID string) (*fcpb.SnapshotKeys, error) {
+	pd, err := digest.ComputeForMessage(task.GetExecutionTask().GetCommand().GetPlatform(), repb.DigestFunction_SHA256)
```
Comment (tylerwilliams) [inline enterprise/server/remote_execution/snaploader/snaploader.go:52]: kind of a nit, but when I see lots of deep references like this (especially repeated ones), I'll often pull define the common bit like:

```
func SnapshotKeys(scheduled *repb.ScheduledTask, ...

task := scheduled.GetExecutionTask()
```

then you can shorten this to `task.GetCommand().GetPlatform()` and below to `task.GetExecuteRequest().GetInstanceName()`

File: enterprise/server/remote_execution/containers/firecracker/firecracker.go [hunk 5]
```diff
@@ -500,20 +500,20 @@ func NewContainer(ctx context.Context, env environment.Env, task *repb.Execution
...
-		recyclingEnabled := platform.IsTrue(platform.FindValue(task.GetCommand().GetPlatform(), platform.RecycleRunnerPropertyName))
+		recyclingEnabled := platform.IsTrue(platform.FindValue(task.GetExecutionTask().GetCommand().GetPlatform(), platform.RecycleRunnerPropertyName))
 		if recyclingEnabled && *snaputil.EnableLocalSnapshotSharing {
-			_, err := loader.GetSnapshot(ctx, c.snapshotKey)
+			_, err := loader.GetSnapshot(ctx, c.snapshotKeys)
 			c.createFromSnapshot = (err == nil)
 		}
 	} else {
-		c.snapshotKey = opts.SavedState.GetSnapshotKey()
+		c.snapshotKeys = &fcpb.SnapshotKeys{MainKey: opts.SavedState.GetSnapshotKey()}
```
Comment (tylerwilliams) [inline enterprise/server/remote_execution/containers/firecracker/firecracker.go:516]: not sure I have a better name but I found this confusing that snapshotKey**s** is a single proto rather than a slice of keys.

snapshotKeySet or SnapshotIdentifier or something? idk

**Review by tylerwilliams — CHANGES_REQUESTED** [review]

File: enterprise/server/remote_execution/snaploader/snaploader.go (hunk 2 above)
Comment (bduffany) [reply enterprise/server/remote_execution/snaploader/snaploader.go:223]: I considered that, but kind of like having a separation between the keys that are read+write vs. the keys that are readonly.

File: enterprise/server/remote_execution/containers/firecracker/firecracker.go (hunk 5 above)
Comment (bduffany) [reply enterprise/server/remote_execution/containers/firecracker/firecracker.go]: Yeah this did feel mildly awkward. I forgot to come back around and rename it. SnapshotKeySet sgtm - done

File: enterprise/server/remote_execution/containers/firecracker/firecracker_test.go (hunk 1 above)
Comment (bduffany) [reply enterprise/server/remote_execution/containers/firecracker/firecracker_test.go:2361]: woops, fixed

File: enterprise/server/remote_execution/snaploader/snaploader.go (hunk 4 above)
Comment (bduffany) [reply enterprise/server/remote_execution/snaploader/snaploader.go]: done

File: enterprise/server/remote_execution/snaploader/snaploader.go (hunk 3 above)
Comment (bduffany) [reply enterprise/server/remote_execution/snaploader/snaploader.go]: woops, reverted this readKey vs writeKey stuff. This key is only used in UnpackSnapshot so it's already readonly and it's the one associated w/ the manifest.

File: enterprise/server/remote_execution/containers/firecracker/firecracker.go [hunk 6]
```diff
@@ -737,7 +737,11 @@ func (c *FirecrackerContainer) saveSnapshot(ctx context.Context, snapshotDetails
 	}
 
 	snaploaderStart := time.Now()
-	if err := c.loader.CacheSnapshot(ctx, c.snapshotKey, opts); err != nil {
+	// Note: we only update the main snapshot here, i.e. the one corresponding
+	// to the pushed git branch. We do not update the fallback key(s) that we
+	// might've read from, i.e. the ones corresponding to the PR base branch or
+	// repo's default branch.
```
Comment (tylerwilliams) [inline enterprise/server/remote_execution/containers/firecracker/firecracker.go:743]: // NB: we only update the snapshot corresponding to the pushed git branch.
	// We do not update the snapshot for any fallback key(s) that we may have
	// read from, i.e. ones corresponding to the PR's base branch or the
	// repo's default branch.

File: enterprise/server/remote_execution/snaploader/snaploader.go [hunk 7]
```diff
@@ -693,7 +756,7 @@ func UnpackContainerImage(ctx context.Context, l *FileCacheLoader, imageRef, ima
 	// ChunkedFile then add it to cache.
 	// TODO(bduffany): single-flight this.
 	start := time.Now()
-	cow, err := copy_on_write.ConvertFileToCOW(ctx, l.env, imageExt4Path, chunkSize, outDir, key.InstanceName)
+	cow, err := copy_on_write.ConvertFileToCOW(ctx, l.env, imageExt4Path, chunkSize, outDir, "" /*=instanceName*/)
```
Comment (tylerwilliams) [inline enterprise/server/remote_execution/snaploader/snaploader.go:759]: why the removal of instance name here?

**Review by tylerwilliams — CHANGES_REQUESTED** [review]

File: enterprise/server/remote_execution/snaploader/snaploader.go [hunk 8]
```diff
@@ -42,21 +43,58 @@ const (
...
+		keys.FallbackKeys = append(keys.FallbackKeys, fallbackKey)
+	}
+	return keys, nil
+}
+
+func gitRefs(task *repb.ExecutionTask) (branchRef string, fallbackRefs []string) {
+	// NOTE: keep these names in sync with workflow service
+	branchRef = getEnv(task, "GIT_BRANCH")
+	if branchRef == "" {
+		return "", nil
```
Comment (maggie-lou) [inline enterprise/server/remote_execution/snaploader/snaploader.go:80]: What does this case mean? Should we either return an error that the env vars weren't set correctly, or use the fallbacks if they're set?

File: enterprise/server/remote_execution/snaploader/snaploader_test.go [hunk 9]
```diff
@@ -78,20 +78,82 @@ func TestPackAndUnpackChunkedFiles(t *testing.T) {
...
+	require.Equal(t, "main", forkKeys.FallbackKeys[1].Ref)
+
+	// Sanity check that the branch key is not found in cache, to make sure
+	// we're actually falling back.
+	_, err = loader.GetSnapshot(ctx, &fcpb.SnapshotKeySet{BranchKey: forkKeys.GetBranchKey()})
+	require.Error(t, err)
+
+	// Now try unpacking with the complete PR key set, including fallbacks. This
+	// should succeed.
+	mustUnpack(t, ctx, loader, forkKeys, forkWorkDir, opts)
```
Comment (maggie-lou) [inline enterprise/server/remote_execution/snaploader/snaploader_test.go:154]: Could you add a test that we don't try to update any of the fallback snapshots when we save the snapshot?

File: proto/firecracker.proto [hunk 10]
```diff
@@ -26,6 +26,20 @@ message VMConfiguration {
   // TODO: add container_image here?
 }
 
+message SnapshotKeySet {
+  // Snapshot key to be read from when resuming the snapshot and written to when
+  // the task is complete.
+  //
+  // For workflows, this is associated with the git branch checked out in the
+  // snapshotted git repo directory.
+  SnapshotKey branch_key = 1;
```
Comment (maggie-lou) [inline proto/firecracker.proto:35]: Personally I prefer the language of something like `primary_key`. It's harder for me to conceptualize what a `branch_key` is

File: enterprise/server/remote_execution/snaploader/snaploader.go (hunk 7 above)
Comment (bduffany) [reply enterprise/server/remote_execution/snaploader/snaploader.go:759]: Our current "non-COW" OCI -> ext4 conversion does not have instance name in the cache key so I think it was unintentional when I introduced instance name into the cache key originally.

Do you think we should include the instance name?

File: enterprise/server/remote_execution/containers/firecracker/firecracker.go (hunk 6 above)
Comment (bduffany) [reply enterprise/server/remote_execution/containers/firecracker/firecracker.go]: Done

File: enterprise/server/remote_execution/snaploader/snaploader.go (hunk 8 above)
Comment (bduffany) [reply enterprise/server/remote_execution/snaploader/snaploader.go:83]: I figured if `GIT_BRANCH` is not set, then we're not looking at a well-formed workflow action, so to be on the safe side we should ignore the fallbacks. Added a clarifying comment.

File: enterprise/server/remote_execution/snaploader/snaploader_test.go (hunk 9 above)
Comment (bduffany) [reply enterprise/server/remote_execution/snaploader/snaploader_test.go:154]: `CacheSnapshot` only accepts a single key, not a `SnapshotKeySet` so I think it's not something that can be tested in this package.

We could add a firecracker_test case but it felt a bit overkill.

File: proto/firecracker.proto (hunk 10 above)
Comment (bduffany) [reply proto/firecracker.proto:35]: haha, I originally discussed this with Tyler offline and we found `main_key` / `primary_key` to also be confusing and settled on `branch_key` :sweat_smile: 

I think this name is not super consequential so would prefer to leave as-is for now.

File: enterprise/server/remote_execution/snaploader/snaploader.go (hunk 8 above)
Comment (tylerwilliams) [reply enterprise/server/remote_execution/snaploader/snaploader.go:87]: This scares me a little -- are only workflows is allowed to use snapshotting? What about recyclable runners with container-isolation-type=firecracker?

File: enterprise/server/remote_execution/snaploader/snaploader.go (hunk 7 above)
Comment (tylerwilliams) [reply enterprise/server/remote_execution/snaploader/snaploader.go:766]: Maybe add a comment that htis is nil to match OCI->ext4 conversion?

**Review by tylerwilliams — APPROVED** [review]
LGTM but please make sure @maggie-lou is good before submitting

File: enterprise/server/remote_execution/snaploader/snaploader.go (hunk 8 above)
Comment (bduffany) [reply enterprise/server/remote_execution/snaploader/snaploader.go:87]: Those will still be allowed to use snapshotting but just won't have a git ref as part of the snapshot key.

Comment (maggie-lou) [reply enterprise/server/remote_execution/snaploader/snaploader.go:87]: Is there a reason we don't want to include git ref for these non-workflow tasks? Seems like it would improve their snapshot match quality

**Review by maggie-lou — APPROVED** [review]
Had one question about including git ref for non-workflow tasks. Otherwise LGTM

File: enterprise/server/remote_execution/snaploader/snaploader.go (hunk 8 above)
Comment (bduffany) [reply enterprise/server/remote_execution/snaploader/snaploader.go:87]: In general, actions will not have these git ref env vars set. We are setting these specially for workflow actions.

Comment (bduffany) [reply enterprise/server/remote_execution/snaploader/snaploader.go:87]: It's unclear to me whether git ref is the right thing there, vs something like a hash of the Action proto.

File: enterprise/server/remote_execution/snaploader/snaploader.go (hunk 7 above)
Comment (bduffany) [reply enterprise/server/remote_execution/snaploader/snaploader.go]: Done
